runtime/service.go

```diff
@@ -489,6 +490,47 @@ func dialVsock(ctx context.Context, contextID uint32, port uint32) (net.Conn, er
 	return nil, lastErr
 }
 
+// findNextAvailableVsockCID finds first available vsock context ID.
+// It uses VHOST_VSOCK_SET_GUEST_CID ioctl which allows some CID ranges to be statically reserved in advance.
+// The ioctl fails with EADDRINUSE if cid is already taken and with EINVAL if the CID is invalid.
+// Taken from https://bugzilla.redhat.com/show_bug.cgi?id=1291851
+func findNextAvailableVsockCID() (uint32, error) {
+	const (
+		ioctlVsockSetGuestCID = uintptr(0x4008AF60) // Corresponds to VHOST_VSOCK_SET_GUEST_CID in vhost.h
+		startCID              = uint32(3)           // 0, 1 and 2 are reserved CIDs, see http://man7.org/linux/man-pages/man7/vsock.7.html
+		maxCID                = math.MaxUint32
+	)
+
+	file, err := os.OpenFile("/dev/vhost-vsock", syscall.O_RDWR, 0666)
+	if err != nil {
+		return 0, errors.Wrap(err, "failed to open vsock device")
+	}
+
+	defer file.Close()
+
+	for contextID := startCID; contextID < maxCID; contextID++ {
```

Comment:
We may want to provide a context in this function since this could iterate for a very long time